coins/ethereum/contracts/Router.sol

```diff
@@ -0,0 +1,101 @@
+//SPDX-License-Identifier: AGPLv3
+pragma solidity ^0.8.0;
+
+import "./Schnorr.sol";
+
+contract Router is Schnorr {
+    // contract owner
+    address owner;
+
+    // nonce is incremented for each batch of transactions executed
+    uint256 nonce; 
+
+    // prevents re-entrancy
+    uint8 internal locked;
+
+    struct PublicKey {
+        uint8 parity;
+        bytes32 px;
+    }
+
+    // current aggregated validator public key 
+    PublicKey publicKey;
+
+    struct Transaction {
+        address to;
+        uint256 value;
+        uint256 gas;
+        bytes data;
+    }
+
+    struct Signature {
+        bytes32 e;
+        bytes32 s;
+    }
+
+    event Executed(uint256 nonce, uint256 index, bool success);
+
+    constructor() {
+        owner = msg.sender;
+    }
+
+    modifier onlyOwner() {
+        require(msg.sender == owner, "must be called by the contract owner");
+        _;
+    }
+
+    modifier noReentrant() {
+        require(locked == 0, "no re-entrancy");
```

Comment:
This is sub-optimal IIRC. The gas refund from deleting locked is < the gas refund from restoring the original value when the initial set additionally allocates the memory. OZ's ReentrancyGuard discusses this.

Reply:
updated to use OZ ReentrancyGuard